### ISSUE TITLE
Steep Teal Osprey - Re-entrancy through `_isHighRiskProposal` allows attacker to create multiple active proposals from one address

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-01-olympus-on-chain-governance-judging#90]
Steep Teal Osprey

high

# Re-entrancy through `_isHighRiskProposal` allows attacker to create multiple active proposals from one address

## Summary
 
Re-entrancy through `_isHighRiskProposal` allows attacker to create multiple active proposals from one EOA, even though they should only be able to use their voting power / whitelist status to create one active proposal at a time.

## Vulnerability Detail

Each proposer should only be allowed to have one proposal active at a time with their voting power, as observed in the following code in `propose` function:

```solidity
        uint256 latestProposalId = latestProposalIds[msg.sender];
        if (latestProposalId != 0) {
            ProposalState proposersLatestProposalState = state(latestProposalId);
            if (proposersLatestProposalState == ProposalState.Active)
                revert GovernorBravo_Proposal_AlreadyActive();
            if (proposersLatestProposalState == ProposalState.Pending)
                revert GovernorBravo_Proposal_AlreadyPending();
        }
```

However, in `_isHighRiskProposal`, the following code exists:

```solidity
                    if (action == 1) {
                        // Check if the module has a high risk keycode
                        if (isKeycodeHighRisk[Module(actionTarget).KEYCODE()]) return true;
                    }
```

There is no real check to see if `actionTarget`, which is a user provided address, is a real Olympus module, so this call could be to an arbitrary attacker controlled contract. The attacker controlled contract can then reenter and call `propose` once more. The above check on `latestProposalId` will be completely bypassed because the first call has not yet set the `latestProposalId`, so the attacker will be able to create multiple active proposals. 

## Impact

One address can create multiple active proposals. 

## Code Snippet

https://github.com/sherlock-audit/2024-01-olympus-on-chain-governance/blob/main/bophades/src/external/governance/GovernorBravoDelegate.sol#L127

## Tool used

Manual Review

## Recommendation
Use reentrancy modifiers to prevent reentrancy

**Comment by sherlock-admin2** [comment]
> escalate
> This issue is invalid and is not a dup of https://github.com/sherlock-audit/2024-01-olympus-on-chain-governance-judging/issues/62 . It points out that the reentrancy is via `KEYCODE` which is not possible because [`KEYCODE`](https://github.com/sherlock-audit/2024-01-olympus-on-chain-governance/blob/6171681cfeec8a24b0449f988b75908b5e640a35/bophades/src/Kernel.sol#L119) is a pure function so the calls will be done via staticcalls and will not allow to execute attack as described in https://github.com/sherlock-audit/2024-01-olympus-on-chain-governance-judging/issues/62

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
@s1ce Could you provide a PoC to verify your issue, if not I have to agree with @r0ck3tzx.

**Comment by 0xLienid** [comment]
It's basically the same vulnerability even if the reentrancy happens on a different call in a policy/module

**Comment by Czar102** [comment]
I agree with the escalation, the submitter didn't realize how to handle the attack and we can't reward a submission which "by accident" had a very similar path as another valid finding.

Planning to consider this a unique invalid report and accept the escalation.

**Comment by Czar102** [comment]
Result:
Invalid
Unique

**Comment by sherlock-admin** [comment]
Escalations have been resolved successfully!

Escalation status:
- [r0ck3tzx](https://github.com/sherlock-audit/2024-01-olympus-on-chain-governance-judging/issues/90/#issuecomment-1916551676): accepted
